### PR TITLE
PHPLIB-1118: Drop support for PHP 7.2 and 7.3

### PR DESCRIPTION
**Pull request by alcaeus**
PHPLIB-1118

This drops support for PHP 7.2 and 7.3 and updates code for PHP 7.4. Changes have been made automatically using rector and phpcbf. A few manual fixes were applied (a few property types were fixed to exclude `null`).

### CODE REVIEW

File: src/Model/BSONIterator.php
Comment (alcaeus) [inline src/Model/BSONIterator.php:44]: Note that rector keeps changing this to `private ?object $current = null;`, which is incorrect as the return type of `toPHP` depends on the type map provided. @GromNaN is there a way to tell rector about this?

File: examples/bulk.php
Comment (alcaeus) [inline examples/bulk.php:71]: This change is responsible for most of the diff in this PR: a trailing comma after the last argument in multi-line method calls is now required. This is consistent with the rules already in place for multi-line arrays.

File: examples/persistable.php
Comment (alcaeus) [inline examples/persistable.php:19]: I went through the various type changes and confirmed they were accurate, but a second look certainly won't hurt.

File: src/Model/BSONIterator.php
Comment (GromNaN) [reply src/Model/BSONIterator.php:44]: It looks like we are using rector with an outdated version of phpstan stubs. https://github.com/phpstan/phpstan-src/blob/1.10.21/resources/functionMap.php#L6735

Comment (GromNaN) [reply src/Model/BSONIterator.php:44]: To disable the problematic rule, add this to `rector.php`:
```php
    $rectorConfig->skip([
        // ...
        TypedPropertyFromAssignsRector::class => [
            __DIR__ . '/src/Model/BSONIterator.php',
        ],
```

File: src/Collection.php
Comment (GromNaN) [inline src/Collection.php:87]: The type have been missed on this properties.

File: src/Database.php
Comment (GromNaN) [inline src/Database.php:69]: Missing type

File: src/Operation/CreateIndexes.php
Comment (GromNaN) [inline src/Operation/CreateIndexes.php:143]: This function is `strval`.

File: src/Operation/ListCollectionNames.php
Comment (GromNaN) [inline src/Operation/ListCollectionNames.php:78]:
```suggestion
            'strval'
```

File: tests/Collection/CollectionFunctionalTest.php
Comment (GromNaN) [inline tests/Collection/CollectionFunctionalTest.php:188]: Good improvement. Even if that doesn't change anything for the test.

File: tests/UnifiedSpecTests/UnifiedTestRunner.php
Comment (GromNaN) [inline tests/UnifiedSpecTests/UnifiedTestRunner.php:73]: This property is never set. https://github.com/mongodb/mongo-php-library/pull/1129

**Review by GromNaN — APPROVED** [review]
LGTM. With a few optional comments.

File: examples/persistable.php
Comment (GromNaN) [reply examples/persistable.php:19]: Good to me. Some missing property types from `ext-mongodb`, maybe due to outdated stubs.

When types are added to properties, the biggest risk is to have an error when trying to access an uninitialized property.

File: tests/UnifiedSpecTests/UnifiedTestRunner.php
Comment (jmikola) [reply tests/UnifiedSpecTests/UnifiedTestRunner.php:73]: Used by Atlas Testing. See https://github.com/mongodb/mongo-php-library/pull/1129#discussion_r1262872050.

File: src/Operation/ListCollectionNames.php
Comment (jmikola) [reply src/Operation/ListCollectionNames.php:78]: No objection here unless we wanted to introduce a CollectionInfo type hint to the callable instead.

File: src/Operation/Watch.php
Comment (jmikola) [inline src/Operation/Watch.php:79]: These are missing an `array` type.

File: tests/Collection/CollectionFunctionalTest.php
Comment (jmikola) [inline tests/Collection/CollectionFunctionalTest.php:708]: Since this is part of a multi-line function call, shouldn't there be a trailing comma?

Likewise for the diff immediately below.

File: tests/Collection/CrudSpecFunctionalTest.php
Comment (jmikola) [inline tests/Collection/CrudSpecFunctionalTest.php:91]: Noted that `512` is the default value for `$depth`.

File: tests/DocumentationExamplesTest.php
Comment (jmikola) [inline tests/DocumentationExamplesTest.php:1714]: Since this appears within a documentation example, I think we should use a leading backslash.

```suggestion
            echo json_encode($e->getResultDocument(), \JSON_THROW_ON_ERROR);
```

File: tests/Operation/CreateEncryptedCollectionFunctionalTest.php
Comment (jmikola) [inline tests/Operation/CreateEncryptedCollectionFunctionalTest.php:8]: Any reason this wasn't caught before?

File: tests/SpecTests/Operation.php
Comment (jmikola) [inline tests/SpecTests/Operation.php:562]: Do you want to split the `array_map()` parameters across multiple lines for readability? This extends past 120 characters.

File: tests/SpecTests/Operation.php
Comment (jmikola) [inline tests/SpecTests/Operation.php:651]: This could also be `strval` if we don't care about the IndexInfo type-checking.

File: tests/SpecTests/RetryableReadsSpecTest.php
Comment (jmikola) [inline tests/SpecTests/RetryableReadsSpecTest.php:61]: This is necessary because we don't specify an array shape for `$skippedOperations`, correct?

File: tests/SpecTests/RetryableWritesSpecTest.php
Comment (jmikola) [inline tests/SpecTests/RetryableWritesSpecTest.php:110]: Since this is only used within this class, would it make sense to use `self`? This was conveniently [added in PHP 7.4](https://php.watch/versions/7.4/typed-properties).

FunctionalTestCase here refers to SpecTests/FunctionalTestCase, even though this actually only requires the base FunctionalTestCase class.

File: tests/UnifiedSpecTests/RunOnRequirement.php
Comment (jmikola) [inline tests/UnifiedSpecTests/RunOnRequirement.php:40]: Should we have kept `stdClass` here?

File: tests/UnifiedSpecTests/UnifiedTestCase.php
Comment (jmikola) [inline tests/UnifiedSpecTests/UnifiedTestCase.php:28]: Interesting that rector left `stdClass` here but changed it to `object` (I assume) in `RunOnRequirement.php`.

File: tests/Collection/CollectionFunctionalTest.php
Comment (alcaeus) [reply tests/Collection/CollectionFunctionalTest.php:708]: I can only imagine that the sniff gets tripped up by the closure, but I'll have to investigate.

File: tests/Operation/CreateEncryptedCollectionFunctionalTest.php
Comment (alcaeus) [reply tests/Operation/CreateEncryptedCollectionFunctionalTest.php:8]: I think psalm would've caught this, but we haven't enabled it for tests. I'll enable it in PHPLIB-1119 and add all current test errors to the baseline.

File: tests/Collection/CollectionFunctionalTest.php
Comment (alcaeus) [reply tests/Collection/CollectionFunctionalTest.php:708]: Since we're in a new class definition, `self` won't do what we expect it to do, as it will refer to the anonymous class: https://3v4l.org/GavRg

File: tests/Collection/CrudSpecFunctionalTest.php
Comment (alcaeus) [reply tests/Collection/CrudSpecFunctionalTest.php:91]: Yes, I believe that the `assertIsObject` check later in this file tripped rector up a bit.

File: src/Operation/Watch.php
Comment (alcaeus) [reply src/Operation/Watch.php:79]: Correct. I've added a type hint to avoid the cast.

File: tests/DocumentationExamplesTest.php
Comment (GromNaN) [reply tests/DocumentationExamplesTest.php]: For the purpose of documentation, we should not add this option : it complicates the example without adding anything useful. The JSON returned by mongo is always valid. 
The rector rule can be disabled for this file.

Comment (alcaeus) [reply tests/DocumentationExamplesTest.php]: I agree. I excluded the file from that rector.

File: tests/UnifiedSpecTests/RunOnRequirement.php
Comment (alcaeus) [reply tests/UnifiedSpecTests/RunOnRequirement.php]: It got tripped up by a `assertIsObject` check later in the file.

File: src/Operation/ListCollectionNames.php
Comment (alcaeus) [reply src/Operation/ListCollectionNames.php]: I undid that check as this is in the CallbackIterator: the callback is invoked with two arguments, but `strval` only accepts one.

File: tests/SpecTests/RetryableReadsSpecTest.php
Comment (alcaeus) [reply tests/SpecTests/RetryableReadsSpecTest.php]: Correct. I removed this in favour of a proper `array<string, string>` annotation on the property.

File: tests/SpecTests/RetryableWritesSpecTest.php
Comment (alcaeus) [reply tests/SpecTests/RetryableWritesSpecTest.php:110]: Since we're in a new class definition, `self` won't do what we expect it to do, as it will refer to the anonymous class: https://3v4l.org/GavRg

File: tests/DocumentationExamplesTest.php
Comment (jmikola) [reply tests/DocumentationExamplesTest.php]: Ah, I didn't realize rector was adding this. I agree it's not needed for example code.

File: tests/UnifiedSpecTests/EntityMap.php
Comment (jmikola) [inline tests/UnifiedSpecTests/EntityMap.php:38]: This is incorrect. The `set()` method assigns the lsid object with a string key.

```suggestion
     * @var array<string, stdClass>
```

**Review by jmikola — APPROVED** [review]
LGTM after the necessary change in `EntityMap.php`.

Everything else is suggestions, for which I'll defer to you.

File: src/Operation/Aggregate.php
Comment (alcaeus) [reply src/Operation/Aggregate.php:57]: I'd wait until we start using array shapes more, as that change would trigger a number of psalm issues down the line.

File: src/Operation/InsertMany.php
Comment (alcaeus) [reply src/Operation/InsertMany.php:46]: I chose this as it's what the original type was about, but thinking about it made me realise that there's nothing that would prevent a mixed list of arrays and objects to be passed. I've updated it.

File: tests/SpecTests/ErrorExpectation.php
Comment (alcaeus) [reply tests/SpecTests/ErrorExpectation.php:25]: That one was added manually to avoid the cast. I've added a shape here as well.

File: tests/SpecTests/ErrorExpectation.php
Comment (alcaeus) [reply tests/SpecTests/ErrorExpectation.php:29]: Not sure why those weren't addressed, I'll let @GromNaN follow up on this.

File: tests/UnifiedSpecTests/EntityMap.php
Comment (alcaeus) [reply tests/UnifiedSpecTests/EntityMap.php]: Oh, I missed that. Thanks for catching it!

File: tests/SpecTests/RetryableWritesSpecTest.php
Comment (alcaeus) [reply tests/SpecTests/RetryableWritesSpecTest.php]: That's a good point. I've updated it to `RetryableWritesSpecTest`.

File: examples/persistable.php
Comment (alcaeus) [reply examples/persistable.php:19]: Note: the missing property types weren't caught by Rector for some reason, and I only later realised that we had an exclusion in the phpcs config to prevent native types from being added. This has since been removed and types for properties are now enforced.

**Pull request merged**
